### PR TITLE
Add Papago Translation Capabilities

### PR DESCRIPTION
**Pull request by bjrambo**
주석 2020-09-05 201432
It is translations for Papago. 

use to Papago api.

### CODE REVIEW

File: src/DustyBot/Modules/BotModule.cs [hunk 1]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
             await command.ReplySuccess(Communicator, $"Removed {removed} supporters.");
         }
 
+        [Command("번역", "Translate words and sentences.")]
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:357]: To be consistent with other commands, primary usage should be `translate`. Put `번역` and `tr` as aliases please.

File: src/DustyBot/Modules/BotModule.cs [hunk 2]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+        [Alias("tr")]
+        [Parameter("start", ParameterType.String, "Select the language that needs to be translated.")]
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:359]: Capitalize `start`

File: src/DustyBot/Modules/BotModule.cs [hunk 3]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+        [Parameter("end", ParameterType.String, "Select the language to translate.")]
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:360]: Capitalize `end`

File: src/DustyBot/Modules/BotModule.cs [hunk 4]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+        [Parameter("Message", ParameterType.String, ParameterFlags.Remainder, "Input the word or sentence you want to translate.")]
+        public async Task Translation(ICommand command)
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:362]: This doesn't really belong in the `Bot` module. I'm not sure where to put it. Maybe put it into the `Info` module for now? Or create a new `Translator` module.

File: src/DustyBot/Modules/BotModule.cs [hunk 5]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+        {
+            var config = await Settings.ReadGlobal<BotConfig>();
+
+            await command.Message.Channel.TriggerTypingAsync();
+            string stringMessage = command["Message"].ToString();
+            string FirstLang = command["start"].ToString();
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:368]: FirstLang -> firstLang

File: src/DustyBot/Modules/BotModule.cs [hunk 6]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+            string LastLang = command["end"].ToString();
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:369]: LastLang -> lastLang

File: src/DustyBot/Modules/BotModule.cs [hunk 7]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+
+            var data = new Dictionary<string, string>()
+            {
+                {"source", FirstLang},
+                {"target", LastLang},
+                {"text", stringMessage},
+            };
+
+            //TODO(BJRambo) : checking to why did Reload httpclient later.
+            using (var papagoClient = new HttpClient())
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:379]: This is a wrong usage of HttpClient, check https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/
Use `WebRequest.CreateHttp` for now please. I plan to switch to IHttpClientFactory in all commands later.

File: src/DustyBot/Modules/BotModule.cs [hunk 8]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+            {
+                papagoClient.DefaultRequestHeaders.Clear();
+                papagoClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+                papagoClient.DefaultRequestHeaders.TryAddWithoutValidation("Content-Type", "application/json");
+
+                using (var request = new HttpRequestMessage(HttpMethod.Post, "https://openapi.naver.com/v1/papago/n2mt") { Content = new FormUrlEncodedContent(data) })
+                {
+                    request.Headers.Add("X-Naver-Client-Id", config.PapagoClientId);
+                    request.Headers.Add("X-Naver-Client-Secret", config.PapagoClientSecret);
+                    var responseClinet = await papagoClient.SendAsync(request).ConfigureAwait(false);
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:389]: `ConfigureAwait(false)` not necessary

Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:389]: `responseClinet` -> `responseClient`

File: src/DustyBot/Modules/BotModule.cs [hunk 9]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+                        }
+
+                        EmbedBuilder embedBuilder = new EmbedBuilder()
+                        {
+                            Title = $"Translate from **{FirstLang.ToUpper()}** to **{LastLang.ToUpper()}**"
+                        };
+
+                        embedBuilder.WithDescription(translateSentence);
+                        embedBuilder.WithColor(new Color(0, 206, 56));
+                        embedBuilder.WithFooter("powerd by papago.naver.com");
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:408]: `Powered by Papago`

**Comment by bjrambo** [comment]
Korean = ko
Japan = ja
English = en 
Chinese(Simplified) = zh-CH
Chinese(Traditional) = zh-TW
Spanish = es
French = fr
German = de
Russian = ru
Portuguese = pt
Italian = it
Vietnamese = vi
Thai = th
Indonesian = id

File: src/DustyBot/Modules/BotModule.cs [hunk 10]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+                    request.Headers.Add("X-Naver-Client-Id", config.PapagoClientId);
+                    request.Headers.Add("X-Naver-Client-Secret", config.PapagoClientSecret);
+                    var responseClinet = await papagoClient.SendAsync(request).ConfigureAwait(false);
+                    if (responseClinet.IsSuccessStatusCode)
+                    {
+                        var ParserObject = JObject.Parse(await responseClinet.Content.ReadAsStringAsync());
+                        string trMessage = ParserObject["message"]["result"]["translatedText"].ToString();
+
+                        string translateSentence = trMessage;
+                        if (translateSentence.Length > EmbedBuilder.MaxDescriptionLength)
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:396]: Use `Truncate` from `DustyBot.Framework.Utility.StringExtensions`:
`var translateSentence = trMessage.Truncate(EmbedBuilder.MaxDescriptionLength);`

File: src/DustyBot/Modules/BotModule.cs [hunk 11]
```diff
@@ -351,6 +354,64 @@ public async Task RemoveSupporter(ICommand command)
...
+        [Alias("tr")]
+        [Parameter("start", ParameterType.String, "Select the language that needs to be translated.")]
+        [Parameter("end", ParameterType.String, "Select the language to translate.")]
+        [Parameter("Message", ParameterType.String, ParameterFlags.Remainder, "Input the word or sentence you want to translate.")]
+        public async Task Translation(ICommand command)
+        {
+            var config = await Settings.ReadGlobal<BotConfig>();
+
+            await command.Message.Channel.TriggerTypingAsync();
+            string stringMessage = command["Message"].ToString();
```
Comment (yebafan) [inline src/DustyBot/Modules/BotModule.cs:367]: Use `var` everywhere please.

**Review by yebafan — CHANGES_REQUESTED** [review]
Good stuff. I added mostly minor comments, but I try to follow a coding style.

**Comment by yebafan** [comment]
> Korean = ko
> Japan = ja
> English = en
> Chinese(Simplified) = zh-CH
> Chinese(Traditional) = zh-TW
> Spanish = es
> French = fr
> German = de
> Russian = ru
> Portuguese = pt
> Italian = it
> Vietnamese = vi
> Thai = th
> Indonesian = id

Can you add this in the `[Comment("...")]` attribute for the command?

File: src/DustyBot/Modules/TranslatorModule.cs [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+using DustyBot.Framework.Utility;
+
+namespace DustyBot.Modules
+{
+    [Module("Translator", "The ability to translate.")]
+    class TranslatorModule : Module
+    {
+        private ILogger Logger;
```
Comment (yebafan) [inline src/DustyBot/Modules/TranslatorModule.cs:20]: Make this a property (add `{ get; }`)

File: src/DustyBot/Modules/TranslatorModule.cs [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+        public TranslatorModule(ICommunicator communicator, ISettingsProvider settings, ILogger logger)
+        {
+            Communicator = communicator;
+            Settings = settings;
+            //Now not use to it. but it will be use to later.
+            Logger = logger;
+        }
+
+        [Command("tr", "translate")]
```
Comment (yebafan) [inline src/DustyBot/Modules/TranslatorModule.cs:32]: I meant more something like this:
```
[Command("translate", "Translates a piece of text.")]
[Alias("tr"), Alias("번역")]
```

File: src/DustyBot/Modules/TranslatorModule.cs [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+            var data = new Dictionary<string, string>()
+            {
+                {"source", firstLang},
+                {"target", lastLang},
+                {"text", stringMessage},
+            };
+
+            //TODO(BJRambo) : checking to why did Reload httpclient later.
+            var papagoClient = new HttpClient();
```
Comment (yebafan) [inline src/DustyBot/Modules/TranslatorModule.cs:55]: Use `WebRequest.CreateHttp`. This is not how an HttpClient should be used.

**Comment by bjrambo** [comment]
it use, request

10000/day

**Review by yebafan — CHANGES_REQUESTED** [review]

File: src/DustyBot/Modules/TranslatorModule.cs (hunk 13 above)
Comment (bjrambo) [reply src/DustyBot/Modules/TranslatorModule.cs:32]: oh .. sorry.

**Review by yebafan — APPROVED** [review]

**Pull request merged**